flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java

```diff
@@ -0,0 +1,224 @@
+        this.tieredMemorySpecs = new HashMap<>();
+        this.numRequestedBuffers = new AtomicInteger(0);
+        this.bufferFlushCallbacks = new ArrayList<>();
+        this.isRunning = false;
+    }
+
+    @Override
+    public void setup(BufferPool bufferPool, List<TieredStorageMemorySpec> storageMemorySpecs) {
+        this.bufferPool = bufferPool;
+        for (TieredStorageMemorySpec memorySpec : storageMemorySpecs) {
+            checkState(
+                    !tieredMemorySpecs.containsKey(memorySpec.getOwner()),
+                    "Duplicated memory spec.");
+            tieredMemorySpecs.put(memorySpec.getOwner(), memorySpec);
+            numTotalGuaranteedBuffers += memorySpec.getNumGuaranteedBuffers();
+        }
+
+        if (shouldStartBufferFlushChecker) {
+            this.executor =
+                    Executors.newSingleThreadScheduledExecutor(
+                            new ThreadFactoryBuilder()
+                                    .setNameFormat("buffer flush checker")
+                                    .setUncaughtExceptionHandler(FatalExitExceptionHandler.INSTANCE)
+                                    .build());
+            this.executor.scheduleWithFixedDelay(
+                    this::checkShouldFlushCachedBuffers,
+                    DEFAULT_CHECK_FLUSH_INITIAL_DELAY_MS,
+                    DEFAULT_CHECK_FLUSH_PERIOD_DURATION_MS,
+                    TimeUnit.MILLISECONDS);
+        }
+
+        this.isRunning = true;
+    }
+
+    @Override
+    public void registerBufferFlushCallback(Runnable userBufferFlushCallBack) {
+        bufferFlushCallbacks.add(userBufferFlushCallBack);
+    }
+
+    /**
+     * Request a {@link BufferBuilder} instance from {@link BufferPool} for a specific owner. The
+     * {@link TieredStorageMemoryManagerImpl} will not check whether a buffer can be requested and
+     * only record the total number of requested buffers. If the buffers in the {@link BufferPool}
+     * is not enough, this will trigger each tiered storage to flush buffers as much as possible.
+     *
+     * @return the requested buffer
+     */
+    @Override
+    public BufferBuilder requestBufferBlocking() {
+        checkIsRunning();
+
+        MemorySegment requestedBuffer = null;
+        try {
+            requestedBuffer = bufferPool.requestMemorySegmentBlocking();
+        } catch (Throwable throwable) {
+            ExceptionUtils.rethrow(throwable, "Failed to request memory segments.");
+        }
+        numRequestedBuffers.incrementAndGet();
+        checkShouldFlushCachedBuffers();
+        return new BufferBuilder(checkNotNull(requestedBuffer), this::recycleBuffer);
+    }
+
+    @Override
+    public int getMaxAllowedBuffers(Object owner) {
+        checkIsRunning();
+
+        TieredStorageMemorySpec ownerMemorySpec = checkNotNull(tieredMemorySpecs.get(owner));
+        if (ownerMemorySpec.isMemoryReleasable()) {
+            return Integer.MAX_VALUE;
+        } else {
```

Comment:
We may not need the releasable / unreleasable at all.

Reply:
Fixed.